server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+2) If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.
+
+
+## Terms
+
+- Consensus Node (CN): A node that produces and provides blocks.
+- Block Items: The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.
+- Block Hash: A cryptographic hash representing the block’s integrity.
+- Signature: The cryptographic signature of the block hash created by Network private key (aka: LedgerId)
+- Public Key: The public key of the LedgerId that signed the block.
+
+
+## Entities
+
+- **VerificationHandler** 
+  - Receives the stream of block items in the form of List<BlockItemsUnparsed> from the unparsed and unverified block items ring buffer. 
+  - When it detects a block_header, it creates a BlockHashingSession using the BlockHashingSessionFactory, providing it with the initial block items (and internally, the session will handle asynchronous hashing).
+  - Adds subsequent block items to the session, including the block_proof.
+  - Once the block_proof is received, calls completeHashing() on the BlockHashingSession, then releases the thread back to the UnverifiedRingBuffer.
+  - Does not block waiting for verification; the hash computation and verification continue asynchronously.
+- **BlockHashingSessionFactory**
+  - Creates new BlockHashingSession instances, provides them with a ExecutorService.
+- **BlockHashingSession**
+  - Holds all necessary block data.
+  - Accepts block items incrementally. (continues to compute them asynchronously)
+  - Once the block_proof is provided, finalizes the hash computation asynchronously.
+  - After computing the final hash, calls SignatureVerifier for verification.
+- **SignatureVerifier**
+  - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).
+  - Calls BlockStatusManager with the verification result.
+- **BlockStatusManager**
+  - Receives verification results from SignatureVerifier.
+  - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.
+
+## Design
+
+1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.
+2. When the block_header is detected, the `VerificationHandler` creates a `BlockHashingSession` using the `BlockHashingSessionFactory`.
+3. The `BlockHashingSession` accepts subsequent block items incrementally.
+4. Once the block_proof is received, the `BlockHashingSession` calls `completeHashing()` to finalize the hash computation.
+5. Upon completion of computing the final block hash, the `BlockHashingSession` calls the `SignatureVerifier` to verify the signature.
+6. The `SignatureVerifier` compares the computed hash to the hash implied by the signature using the public key.
+7. If the verification fails, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as SIGNATURE_INVALID.
```

Comment:
I also thought about this "ideas" the memory of keeping previous hashes, double check with our own `state` Merkle tree, however, in the sake of KISS decided to leverage the statement that **all blocks are self contained** and keep the verification of each individual block as stateless as possible, by design. 

Adding such double checks are a nice addition, but definitively not needed at a Phase1 implementation of the BlockVerification feature, so maybe we can create a follow-up Epic for BlockVerification Phase2 tasks and add these ideas in there.

Reply:
I agree that such enhancements should be a separate effort.
I disagree that we should not describe high level plans for additional work in this document; there's no reason. that I know of, to restrict the design work here to _only_ what will be competed in the current epic.

Reply:
Yes, I agree that a design document can serve both as a roadmap and/or a blueprint, but at this level for this project, we have been doing the latter. Using them as a tool for socializing implementation.

Reply:
"We have been doing" is not convincing, to me.  We may need to discuss in the team, but I believe that there is little or no value to writing these documents _after_ writing the detail design (we already have full detail, what do we need high-level for at that point?).
Side note, code isn't implementation, it's just detailed design.  Implementation is what the build system produces (automatically).

My opinion (which is not controlling; the team may not agree, at which point we'll document the team decision and commit to that):
Design documents should be where we discuss and debate the high-level design approach for a proposed component before doing the detail design work documented in code files.  This would be where we determine if we have appropriate interfaces, the sequence of events is correct, the core data structures are reasonable, the algorithmic designs are effective, etc...
Documenting a consensus on that high-level design before doing the detail work is valuable enough to warrant a document.  I would contend, however, that "socializing" a design after the fact is much less valuable, and might not be worth the cost.